dashboard/modules/actor/actor_head.py

```diff
@@ -153,39 +138,53 @@ def process_actor_data_from_pubsub(actor_id, actor_table_data):
             # If actor is not new registered but updated, we only update
             # states related fields.
             if actor_table_data["state"] != "DEPENDENCIES_UNREADY":
-                actor_table_data_copy = dict(DataSource.actors[actor_id])
+                actors = DataSource.actors[actor_id]
                 for k in state_keys:
-                    actor_table_data_copy[k] = actor_table_data[k]
-                actor_table_data = actor_table_data_copy
+                    actors[k] = actor_table_data[k]
+                actor_table_data = actors
             actor_id = actor_table_data["actorId"]
-            job_id = actor_table_data["jobId"]
             node_id = actor_table_data["address"]["rayletId"]
             if actor_table_data["state"] == "DEAD":
                 self.dead_actors_queue.append(actor_id)
             # Update actors.
             DataSource.actors[actor_id] = actor_table_data
             # Update node actors (only when node_id is not Nil).
             if node_id != actor_consts.NIL_NODE_ID:
-                node_actors = dict(DataSource.node_actors.get(node_id, {}))
+                node_actors = DataSource.node_actors.get(node_id, {})
                 node_actors[actor_id] = actor_table_data
                 DataSource.node_actors[node_id] = node_actors
-            # Update job actors.
-            job_actors = dict(DataSource.job_actors.get(job_id, {}))
-            job_actors[actor_id] = actor_table_data
-            DataSource.job_actors[job_id] = job_actors
 
         # Receive actors from channel.
         gcs_addr = self._dashboard_head.gcs_address
         subscriber = GcsAioActorSubscriber(address=gcs_addr)
         await subscriber.subscribe()
 
+        total_published_events = 0
+        acc = 0
         while True:
             try:
-                actor_id, actor_table_data = await subscriber.poll()
-                if actor_id is not None:
-                    # Convert to lower case hex ID.
-                    actor_id = actor_id.hex()
-                    process_actor_data_from_pubsub(actor_id, actor_table_data)
+                published = await subscriber.poll(batch_size=1000)
+                start = time.monotonic()
+                for actor_id, actor_table_data in published:
+                    if actor_id is not None:
+                        # Convert to lower case hex ID.
+                        actor_id = actor_id.hex()
+                        process_actor_data_from_pubsub(actor_id, actor_table_data)
+                    total_published_events += 1
+                    # Backpressure the # of actor event process per second.
+                    if total_published_events % ACTORS_TO_PROCESS_PER_SECOND == 0:
+                        await asyncio.sleep(1)
```

Comment:
Could we just yield the loop without sleeping, or sleep a smaller amount?

Reply:
We can also adjust `ACTORS_TO_PROCESS_PER_SECOND`? What's the desired number you think? (rn, it can do up to 1000 events/s)

Reply:
We don't need to throttle this right? The only problem is if it starves other events from being processed.

Actually, I'd favor just removing this limit, since the number of actor events is finite and we yield at some point right? A couple seconds extra of dashboard latency right when starting a lot of actors is probably no big deal, or am I missing something?

Reply:
Let's talk about this in 1:1 tomorrow offline? I prefer to throttle it unless you are really strongly against it because this makes a really huge difference in user experience (I can demonstrate this later through videos). I have 3 reasons why I prefer to throttle it.

1. In small-scale clusters, this has no impact. 1-second delay after processing 1000 events is like nothing in small scale cluster.
2. This just doesn't starve other events. It starves all the API responses, and all the API (including node + job) takes about 5~10 seconds to reply. With throttling, they pretty much respond within 1 second at the same scale.
3. Since the dashboard is a user-facing application, we should have a higher standard for API response time. 5~10 seconds response time for all APIs is not acceptable. Everyone's going to just think the dashboard. 

I think the only con is the actor information can be stale (at very large scale), but it is much better than the whole dashboard being unresponsive. 

I can also disable it after I optimize this processing path as I talked about in the slack channel, and if the response is time is much more reasonable.

Reply:
Addressed based on our discussion (sleep as much as processing time). I found after applying this policy, the load on the dashboard is really good